### PR TITLE
Port MB and Scratch 3.0 to Sugar Desktop and Fix sugar-web

### PR DESCRIPTION
**Pull request by iqraceme**
The focus of this project idea is to fix Sugar Web, get Music blocks and Scratch working in Sugar-Web

### CODE REVIEW

**Comment by walterbender** [comment]
Can you change the title to better reflect the project?

**Comment by samswag** [comment]
@quozl What you are currently asking from this idea is what the student usually come up with on their proposal. They create a timeline, propose more ideas how the challenge can be approached. What @iqraceme and I have done here is to bring up a clear idea of what needs to fix just like the existing merged ideas. 

So far the idea is to fix sugar-web, port MB and Scratch to Sugar which including testing existing web activities and making sure we have a working stable fix and port.

@walterbender maybe I might be wrong on how timeline for GSoC works.

**Comment by pro-panda** [comment]
Thanks, reviewed.
@iqraceme please see https://github.com/sugarlabs/GSoC/pull/6#commits-pushed-e7a6e1d and apply them here as well

Please follow https://github.com/sugarlabs/GSoC/pull/6#issuecomment-455941213 to derive a duration for the completion of this project. We should only put out ideas that we are certain, will justify the timeline of 12 weeks

**Comment by pro-panda** [comment]
Missing link to issue, #5

**Comment by samswag** [comment]
> @quozl What you are currently asking from this idea is what the student usually come up with on their proposal. They create a timeline, propose more ideas how the challenge can be approached. What @iqraceme and I have done here is to bring up a clear idea of what needs to fix just like the existing merged ideas.
> 
> So far the idea is to fix sugar-web, port MB and Scratch to Sugar which including testing existing web activities and making sure we have a working stable fix and port.
> 
> @walterbender maybe I might be wrong on how timeline for GSoC works.

@pro-panda Maybe you need to read this statement before asking @iqraceme for additional work.

**Comment by samswag** [comment]
> Thanks, reviewed.
> @iqraceme please see [#6 (commits)](https://github.com/sugarlabs/GSoC/pull/6#commits-pushed-e7a6e1d) and apply them here as well
> 
> Please follow [#6 (comment)](https://github.com/sugarlabs/GSoC/pull/6#issuecomment-455941213) to derive a duration for the completion of this project. We should only put out ideas that we are certain, will justify the timeline of 12 weeks

What do you mean by "We should only put out ideas that we are certain, will justify the timeline of the 12 weeks"? what is your definition of a project that is 12 weeks worth?  

No current ideas follows #6 and I don't see why this one should.

File: Ideas-2019.md [hunk 1]
```diff
@@ -442,7 +442,7 @@ As with other Sugarizer activities, the new Write activity should be written usi
 
 ------------
 
-## Port Scratch 3.0 to Sugar Desktop and Fix sugar-web
+## Port MB & Scratch 3.0 to Sugar Desktop and Fix sugar-web
```
Comment (walterbender) [inline Ideas-2019.md]: To me, the biggest issue is fixing Sugar Web and making Sugarizer activities once again compatible with Sugar.
MB and Scratch are secondary goals.
With any remaining time, all of the existing Sugar Web/Sugarizer apps could be reviewed and ensured to work.
Any remaining time after that could be added some new apps... I have lots of ideas for those.

Comment (walterbender) [reply Ideas-2019.md]: So, ## Fix Sugar Web, port Scratch 3.0 and Music Blocks, and ensure other web activities are working

**Comment by walterbender** [comment]
A couple of comments:
(1) While I wholeheartedly agree that we need mentors capable of mentoring the coding aspects of our GSoC projects and that coding must be at the heart of the projects, there is more to it than that: we need testing, documentation, etc. I agree that Samson has not recently demonstrated coding skills necessary for mentoring, I think he has demonstrated some of these other skills. That said, I am thrilled that James has agreed to also participate as a mentor.
(2) As I have tried to stress in my previous comments, I think this project is much more about getting our Sugar Web ducks in a line than it is specifically about the Scratch 3.0 port. We have has a number of divergences in approach -- local storage and collaboration in regards to Sugarizer and toolbar management in regards to Music Blocks. There are undoubtedly other incompatibilities as well that should be part of the project scope.

**Comment by walterbender** [comment]
I think another aspect of the project should involve testing Sugar Web across various platforms: I know that a large part of my headaches dealing with JS have been dealing with the vagaries of different browsers and operating systems. While the goal is to get Sugar Web activities running in Sugar, there is demonstrable value in having these activities be able to also run stand-alone in the browser.

**Comment by samswag** [comment]
Thanks for the explanation. it seems like a lot has changed on Gsoc in 2019 that I don't know about. I don't want to argue either about if I still have the skills on working on this project as you stated. Currently, I am currently focused on SL to make sure the tools created are usable. So if what I am doing now will cost me to be ineligible as a mentor, I will gladly step aside. 

The reason I showed interest as a mentor was because I want to see and help this projects work as I find it important. I don't see how will stop me from becoming a GSoC mentor, but I don't want to argue about why I need to be. 

Moreover, given my experience as a GSoC mentor for 2-3 years now, I think that the current Idea that includes sugar-web fix MB and Scratch Port is enough for 3 months of coding. Asking for a timeline is work on the mentor but for the students. This is just a "project Idea" for the student to make a proposal on timeline and approach. 

And as for @iqraceme she was a student just like @pro-panda so working with her I am certain that can mentor this project. I am giving my spot for her to be considered as a mentored.

**Comment by quozl** [comment]
@iqraceme, in light of discussion and changes in https://github.com/sugarlabs/GSoC/pull/29 and in expectation that it will merge, you might rebase from that branch, apply the changes I suggested in https://github.com/sugarlabs/GSoC/pull/6 and put @samswag in as an assisting mentor.  Thanks.

**Comment by pro-panda** [comment]
@samswag wrote,
> Thanks for the explanation. it seems like a lot has changed on Gsoc in 2019 that I don't know about. I don't want to argue either about if I still have the skills on working on this project as you stated. Currently, I am currently focused on SL to make sure the tools created are usable. So if what I am doing now will cost me to be ineligible as a mentor, I will gladly step aside.
The reason I showed interest as a mentor was because I want to see and help this projects work as I find it important. I don't see how will stop me from becoming a GSoC mentor, but I don't want to argue about why I need to be.
Moreover, given my experience as a GSoC mentor for 2-3 years now, I think that the current Idea that includes sugar-web fix MB and Scratch Port is enough for 3 months of coding. Asking for a timeline is work on the mentor but for the students. This is just a "project Idea" for the student to make a proposal on timeline and approach.
And as for @iqraceme she was a student just like @pro-panda so working with her I am certain that can mentor this project. I am giving my spot, for her to be considered as a mentor. As I find her more qualified

Wierd flex, but ok

@quozl wrote,
> @iqraceme, in light of discussion and changes in #29 and in expectation that it will merge, you might rebase from that branch, apply the changes I suggested in #6 and put @samswag in as an assisting mentor. Thanks.

I still disagree with the idea of assisting mentors. See https://github.com/sugarlabs/GSoC/pull/29#issuecomment-460326106

**Comment by GrannieB** [comment]
I came late to this discussion. Here are a few comments on what I see above... re "assisting mentors" and "coding mentors" it seems entirely logical to have these. We had them last summer for Primero (but not officially) as Iqra did all the coding but you probably would have considered me an "assisting mentor." My computer skills stopped with Pascal (I taught AP CS when it used Pascal). Since them IO have dabbled in Python, JavaScript, and HTML and even though I am not "fluent" I do know whether things are sensible to use in a program to make it work... I know what can be done, i.e. what is possible. I also have 30+ years in education so I know what will work with students. Perrie supplied design ideas... her expertise so I guess you could call her an "assisting mentor" as well.

As for requiring a timeline in the PR... last year one of the reasons Iqra's proposal/application did so well was that she proposed a really great timeline. It showed that she had a firm grasp of the task at hand, what skills and resources would be needed and how long everything would take. Her timeline was "spot on" and everything went very well. If we give the students a ready-made  timeline, how will we know they really understand the task and what would be needed to complete it? The answer is simply, we won't.

Finally,  think one of the problems here, which is not new, is that coders often forget that they can't do it all. It takes other people with other skills to effectively bring a software project to a successful and usable place. Assisting mentors can help with this.

**Comment by quozl** [comment]
Thanks for everyone's opinions.

I've had voice calls with @samswag and @iqraceme too.

@walterbender said that our application as a mentoring organisation has been submitted.  So we'll know on 25th February whether we are accepted.

@samswag, @GrannieB, a timeline is not needed yet.  What was needed in previous pull request was enough work to be done to fill the time available.  But Iqra has fixed that already in this pull request by adding more to the requirements.  You can see that in the changes Iqra made.

@iqraceme and I agreed that this time, rather than push my suggested changes to her branch, I'd list the suggested changes that I'd like to see, so that it is clearer which changes are still wanted; the previous pull request has changes that I no longer want.

* [ ] in prerequisites, change "knowledge" to "ability to write in programming languages"; because knowledge is not as required as skill or demonstrated capability,
* [ ] in prerequisites, remove "Strong experience of working on Sugar and activities"; because while it would be helpful it isn't required, it can be learned,
* [ ] in the title, you have "Port MB & Scratch 3.0 to Sugar Desktop and Fix sugar-web", given the changed scope it could be simpler as "Port Scratch activities to Sugar",
* [ ] in the body of the idea, see if you can use any of my green changes in https://github.com/sugarlabs/GSoC/pull/6/commits/0a1b0422cd2afa85311cba710a37475269090c15, especially that Emily's work has since been added to Sugarizer, so it would be better to point the student there rather than to Emily's repository,
* [ ] in the list of steps, add the user media permissions issue, as you can see in my changes in https://github.com/sugarlabs/GSoC/pull/6/commits/57b95ba1a405d8a72b3f7ee0ae88f71a1ee547d1,
* [ ] in the section on mentors, list you and I as coding mentors, and @samswag as one of the assisting mentors; see https://github.com/sugarlabs/GSoC/pull/29 for the discussion about that; though I acknowledge @pro-panda does not consent to that,

Also note that since this pull request was opened, @pro-panda has added Scratch to the new activities idea, in https://github.com/sugarlabs/GSoC/commit/5d22851ae86aa8f898afced171a35cc4a719d379, but there's no reason why adding Scratch can't be in both ideas.

**Comment by walterbender** [comment]
> @walterbender said that our application as a mentoring organisation has been submitted.
We can nonetheless continue to modify/improve out ideas list.

**Comment by iqraceme** [comment]
@Quozl, kindly review. If you think there are still some things needed, I will change it again.

**Comment by quozl** [comment]
@iqraceme, thanks, I've reviewed 03a9f4c, and I approve.  There are some trivial layout changes that could be made, but they'll take a shorter time to do than to describe, so I'll wait.

**Comment by iqraceme** [comment]
@quozl , Will this PR be merged?

**Comment by quozl** [comment]
@iqraceme, not really my decision; I'm in favour of merging, but there are other people who have commented on this pull request, and I'll require their consent to your latest changes; so

* [ ] @walterbender,
* [ ] @samswag,
* [ ] @pro-panda,
* [ ] @GrannieB,

Please click on the "[Files changed](https://github.com/sugarlabs/GSoC/pull/26/files)" tab and review, thanks.

**Comment by samswag** [comment]
Agreed. +1 from me

**Comment by walterbender** [comment]
+1 from me.

**Comment by GrannieB** [comment]
+1 Here

**Comment by pro-panda** [comment]
@quozl Thanks, but I don't know about the scope and utility of the idea, and can't comment on it.

**Comment by quozl** [comment]
Thanks all, merging.

**Pull request merged**